swap/src/network/tor_transport.rs

```diff
@@ -0,0 +1,124 @@
+use data_encoding::BASE32;
+use futures::future::Ready;
+use futures::prelude::*;
+use libp2p::core::multiaddr::{Multiaddr, Protocol};
+use libp2p::core::transport::TransportError;
+use libp2p::core::Transport;
+use libp2p::tcp::tokio::{Tcp, TcpStream};
+use libp2p::tcp::{GenTcpConfig, TcpListenStream, TokioTcpConfig};
+use std::io;
+use std::net::{Ipv4Addr, SocketAddr, SocketAddrV4};
+use std::pin::Pin;
+use tokio_socks::tcp::Socks5Stream;
+use tokio_socks::IntoTargetAddr;
+
+/// Represents the configuration for a TCP/IP transport capability for libp2p.
+#[derive(Clone)]
+pub struct TorTcpConfig {
+    inner: GenTcpConfig<Tcp>,
+    /// Tor SOCKS5 proxy port number.
+    socks_port: u16,
+}
+
+impl TorTcpConfig {
+    pub fn new(tcp: TokioTcpConfig, socks_port: u16) -> Self {
+        Self {
+            inner: tcp,
+            socks_port,
+        }
+    }
+}
+
+impl Transport for TorTcpConfig {
+    type Output = TcpStream;
+    type Error = io::Error;
+    type Listener = TcpListenStream<Tcp>;
+    type ListenerUpgrade = Ready<Result<Self::Output, Self::Error>>;
+    #[allow(clippy::type_complexity)]
+    type Dial = Pin<Box<dyn Future<Output = Result<Self::Output, Self::Error>> + Send>>;
+
+    fn listen_on(self, addr: Multiaddr) -> Result<Self::Listener, TransportError<Self::Error>> {
+        self.inner.listen_on(addr)
+    }
+
+    // dials via Tor's socks5 proxy if configured and if the provided address is an
+    // onion address. or it falls back to Tcp dialling
+    fn dial(self, addr: Multiaddr) -> Result<Self::Dial, TransportError<Self::Error>> {
+        async fn do_tor_dial(socks_port: u16, dest: String) -> Result<TcpStream, io::Error> {
+            tracing::trace!("Connecting through Tor proxy to address: {}", dest);
+            let stream = connect_to_socks_proxy(dest, socks_port)
+                .await
+                .map_err(|e| io::Error::new(io::ErrorKind::ConnectionRefused, e))?;
+            tracing::trace!("Connection through Tor established");
+            Ok(stream)
+        }
+
+        match to_onion_address(addr.clone()) {
+            Some(tor_address_string) => {
+                Ok(Box::pin(do_tor_dial(self.socks_port, tor_address_string)))
+            }
+            _ => self.inner.dial(addr),
+        }
+    }
+
+    fn address_translation(&self, listen: &Multiaddr, observed: &Multiaddr) -> Option<Multiaddr> {
+        self.inner.address_translation(listen, observed)
+    }
+}
+
+/// iterates trhough multi address until we have onion protocol, else return
+/// None Tor expects address in form: ADDR.onion:PORT
+fn to_onion_address(multi: Multiaddr) -> Option<String> {
+    let components = multi.iter();
+    for protocol in components {
+        match protocol {
+            Protocol::Onion(addr, port) => {
+                tracing::warn!("Onion service v2 is being deprecated, consider upgrading to v3");
```

Comment:
They are already deprecated as of now, and support is being removed from the Tor codebase on July 15th. I don't think we should support v2.